diff -Nru usercrond-1.0/crond.pam usercrond-1.0.pam/crond.pam
--- usercrond-1.0/crond.pam	1969-12-31 19:00:00.000000000 -0500
+++ usercrond-1.0.pam/crond.pam	2004-07-26 13:42:16.662309000 -0400
@@ -0,0 +1,12 @@
+#
+# The PAM configuration file for the cron daemon
+#
+#
+auth       required  /lib/security/$ISA/pam_stack.so service=system-auth
+auth       required  pam_env.so
+account    required  /lib/security/$ISA/pam_stack.so service=system-auth
+session    required  /lib/security/$ISA/pam_stack.so service=system-auth
+# Sets up user limits, please uncomment and read /etc/security/limits.conf
+# to enable this functionality.
+# session    required   pam_limits.so
+#
diff -Nru usercrond-1.0/Makefile usercrond-1.0.pam/Makefile
--- usercrond-1.0/Makefile	2004-07-26 14:05:59.719827000 -0400
+++ usercrond-1.0.pam/Makefile	2004-07-26 15:17:04.121157000 -0400
@@ -55,6 +55,7 @@
 DESTSBIN	=	$(DESTROOT)/sbin
 DESTBIN		=	$(DESTROOT)/bin
 DESTMAN		=	$(DESTROOT)/share/man
+DESTETC		=	$(DESTROOT)/../etc
 #<<need bitstring.h>>
 INCLUDE		=	-I.
 #INCLUDE	=
@@ -113,6 +114,7 @@
 install		:	all
 			$(INSTALL) -c -m 755 cron    $(DESTSBIN)/crond
 			$(INSTALL) -c -m 4755 crontab $(DESTBIN)/
+			$(INSTALL) -c -m 0600 crond.pam $(DESTETC)/pam.d/crond
 #			$(INSTALL) -c -m  111 -o root -g crontab -s cron $(DESTSBIN)/
 #			$(INSTALL) -c -m 2111 -o root -g crontab -s crontab $(DESTBIN)/
 			sh putman.sh crontab.1 $(DESTMAN)
