--- usercrond-1.0/cron.c.orig	2007-01-08 14:08:50.000000000 +0000
+++ usercrond-1.0/cron.c	2007-01-08 14:08:43.000000000 +0000
@@ -33,7 +33,7 @@ static	void	usage(void),
 		run_reboot_jobs(cron_db *),
 		find_jobs(int, cron_db *, int, int),
 		set_time(int),
-		cron_sleep(int),
+		cron_sleep(int, cron_db *),
 		sigchld_handler(int),
 		sighup_handler(int),
 		sigchld_reaper(void),
@@ -154,7 +154,7 @@ main(int argc, char *argv[]) {
 
 		/* ... wait for the time (in minutes) to change ... */
 		do {
-			cron_sleep(timeRunning + 1);
+			cron_sleep(timeRunning + 1, &database);
 			set_time(FALSE);
 		} while (clockTime == timeRunning);
 		timeRunning = clockTime;
@@ -258,6 +258,7 @@ main(int argc, char *argv[]) {
 		/* Check to see if we received a signal while running jobs. */
 		if (got_sighup) {
 			got_sighup = 0;
+			database.mtime = (time_t) 0;
 			log_close();
 		}
 		if (got_sigchld) {
@@ -357,7 +358,7 @@ set_time(int initialize) {
  * Try to just hit the next minute.
  */
 static void
-cron_sleep(int target) {
+cron_sleep(int target, cron_db *db) {
 	time_t t1, t2;
 	int seconds_to_wait;
 
@@ -376,6 +377,7 @@ cron_sleep(int target) {
 		 */
 		if (got_sighup) {
 			got_sighup = 0;
+			db->mtime = (time_t) 0;
 			log_close();
 		}
 		if (got_sigchld) {

--- usercrond-1.0/database.c.symlinks	2007-01-28 17:12:24.000000000 +0000
+++ usercrond-1.0/database.c	2007-01-28 17:12:34.000000000 +0000
@@ -354,7 +354,7 @@ static void max_mtime( char *dir_name, s
 		if ( not_a_crontab ( dp ) )
 			continue;
 
-		if (!glue_strings(tabname, sizeof tabname, SPOOL_DIR, dp->d_name, '/'))
+		if (!glue_strings(tabname, sizeof tabname, dir_name, dp->d_name, '/'))
 			continue;	/* XXX log? */
 
 		if ( stat( tabname, &st ) < OK )
