--- usercrond-1.0/security.c.bz181439p2	2006-08-30 08:31:44.000000000 +0200
+++ usercrond-1.0/security.c	2006-08-30 08:31:44.000000000 +0200
@@ -78,17 +78,15 @@
 	return -1;
     }	
 
+#if WITH_SELINUX
     if ( cron_change_selinux_context( u, scontext, file_context ) != 0 )
     {
         syslog(LOG_INFO,"CRON (%s) ERROR: failed to change SELinux context", 
 	       e->pwd->pw_name);
-#if WITH_SELINUX
 	if ( file_context )
 		freecon(file_context);
-#endif
 	return -1;
     }
-#if WITH_SELINUX
     if ( file_context )
 	freecon(file_context);
 #endif
