--- usercrond-1.0/do_command.c.cron_log_facility	2004-08-01 16:12:31.982801000 -0400
+++ usercrond-1.0/do_command.c	2004-08-01 16:12:32.290493000 -0400
@@ -185,6 +185,7 @@
 	PAM_FAIL_CHECK;
 	retcode = pam_setcred(pamh, PAM_ESTABLISH_CRED | PAM_SILENT);
 	PAM_FAIL_CHECK;
+	log_close(); /* PAM has now re-opened our log to auth.info ! */
 #endif
 
 	/* fork again, this time so we can exec the user's command.
